### DISCUSSION TITLE
Building a simple gmail app. Need some guidance.

### DISCUSSION THREAD

**Discussion by surjithctly** [InboxSDK/InboxSDK#814] [Q&A] [answered]
Helllo,

I'm trying to builld a simple gmail extension. So here's what I want to achieve. 

1. From the inbox lists, hover on a conversation and mark it as pinned. 
2. This conversation now move to the top of the inbox. 
3. Clicking again will unpin it and move to its original place. 

How can I use InboxSdK for this simple case? The documentation doesn't explain much. Its too technical. and provides zero example on each methods, properties on how it should be used. 

To achieve the above tasks, can anyone tell which inbox SDK features I should use? and probably how I would be able to pin it (storing its state) and move it to the top (re-order list)?

Thanks in advance. 

~Surjith

**Comment by Macil** [comment, marked as answer]
I think this app idea is a bit redundant with Gmail's own "Starred first" inbox type setting, which moves all starred emails above everything else in your inbox, just like how your pinning feature is described.

But just to explain InboxSDK functionality anyway:

The InboxSDK doesn't have any functionality to make certain emails not render in the inbox. You could make it so pinned emails have the Inbox label removed from them so they don't render normally in the inbox.

Then you could use the ListRouteView's addSection or addCollapsibleSection methods to create a custom DOM element shown at the top in the inbox, which you could then render the email rows into yourself:

```js
InboxSDK.load(1.0, 'route-view-sections-example').then((inboxSDK) => {
  inboxSDK.Router.handleListRoute(
    inboxSDK.Router.NativeRouteIDs.Inbox,
    (listRouteView) => {
      const section = listRouteView.addCollapsibleSection({
        title: 'Demo',
        subtitle: 'subtitle',
        hasDropdown: true,
        onDropdownClick: (event) => {
          event.dropdown.el.textContent = 'Hello world!';
        },
        tableRows: [
          {
            title: 'title',
            body: 'body',
            shortDetailText: 'extra',
            iconUrl: 'https://example.com',
            labels: [
              {
                title: 'label',
                backgroundColor: 'yellow',
                foregroundColor: 'red',
                iconBackgroundColor: 'green',
              },
            ],
            isRead: true,
            onClick: function () {
              console.log('hi');
            },
          },
          {
            title: 'title',
            body: '2222',
            routeID: inboxSDK.Router.NativeRouteIDs.Inbox,
          },
        ],
      });
    }
  );
});
```

(Code sample adapted from one of our test files, [examples/add-route-view-sections/content.js](https://github.com/InboxSDK/InboxSDK/blob/main/examples/add-route-view-sections/content.js))

**Reply by surjithctly** [reply]
Thanks for shedding some light 💡. I agree there is starred email, but if we enabled that we would lose our Primary, Updates tab. So I personally do want a feature like this. I was able to view the collapsible section like I want.

One last question, does inboxSDK support adding custom data/store to an email? For example, I have this code.

```js
ThreadRowView.addButton({
      title: "Pin Email",
      iconUrl: "...",
      onClick: function (event) {
       ...
      },
});
``` 

I want to add a data to the onClick function. eg: `isPinned = true | false`, can I do this with inboxSDK? or should I go with the default `storage.local`? 

Thanks again.

~Surjith

**Reply by Macil** [reply]
No, the InboxSDK doesn't have anything for storing data with an email. (It doesn't have anything for directly storing/modifying data in the user's Gmail account, outside of interacting with the Gmail UI.) You'll need to store that somewhere yourself (like localStorage or on a remote server).

**Reply by surjithctly** [reply]
Thank you. got it. 

btw, the collapsibleSection is missing "description". Usually all mails have a text after the subject right? Seems I cannot add that right now. See image marked with arrow. I see `shortDetailText` but that's right aligned for Date. 

https://inboxsdk.github.io/inboxsdk-docs/router/#rowdescriptor

![CleanShot 2022-12-21 at 13 37 18@2x](https://user-images.githubusercontent.com/1884712/208852845-32a0688a-895a-48e8-bd8a-65a19f672bab.png)
